subgraphs/compound-v3/schema.graphql

```diff
@@ -0,0 +1,2020 @@
+  BORROWER
+}
+
+enum Tranche {
+  " Senior denotes debt with a higher priority. The first debt to be paid back to lenders. "
+  SENIOR
+
+  " Junior tranche denotes lower priority debt. This is secondary priority to be paid back to lenders. "
+  JUNIOR
+}
+
+enum PositionSide {
+  " Position opened as a lender (used as collateral) "
+  COLLATERAL
+
+  " Position opened as a borrower "
+  BORROWER
+}
+
+# Most markets only have a single interest rate given a specific type.
+# However, fixed term lending protocols can have multiple rates with
+# different duration/maturity per market. You can append a counter
+# to the IDs to differentiate.
+type InterestRate @entity {
+  " { Interest rate side }-{ Interest rate type }-{ Market ID }-{ Optional: # days/hours since epoch time } "
+  id: ID!
+
+  " Interest rate in percentage APY. E.g. 5.21% should be stored as 5.21 "
+  rate: BigDecimal!
+
+  " Duration of the loan in days. Only applies to fixed term lending (e.g. Notional) "
+  duration: Int
+
+  " Maturity of the loan in block height. Only applies to fixed term lending (e.g. Notional) "
+  maturityBlock: BigInt
+
+  " The party the interest is paid to / received from "
+  side: InterestRateSide!
+
+  " The type of interest rate (e.g. stable, fixed, variable, etc) "
+  type: InterestRateType!
+
+  " The level of debt priority at this interest rate "
+  tranche: Tranche
+}
+
+enum FeeType {
+  " Fees from liquidations "
+  LIQUIDATION_FEE
+
+  " Fees given to an admin "
+  ADMIN_FEE
+
+  " Fees that are taken by the protocol "
+  PROTOCOL_FEE
+
+  " Fee to mint an asset. Found mostly in CDPs "
+  MINT_FEE
+}
+
+type Fee @entity {
+  " { Fee type }-{ Market ID }-{ Optional: # days/hours since epoch time } "
+  id: ID!
+
+  " Fee in percentage. E.g. 5.21% should be stored as 5.21 "
+  rate: BigDecimal!
+
+  " The type of fee (e.g. liquidation, admin, etc.) "
+  type: FeeType!
+}
+
+enum RevenueSource {
+  " Revenue that is generated from borrower's interest "
+  BORROW_INTEREST
+
+  " Revenue that is generated from liquidation fees "
+  LIQUIDATION_FEE
+
+  " Revenue that is generated from fees from flashloans "
+  FLASHLOAN_FEE
+}
+
+# This entity offers a more nuanced view of revenue types
+# Use this to provide the sources of revenue and amounts of each
+type RevenueDetails @entity {
+  " { Market/Protocol ID }-{ Optional: Snapshot ID } "
+  id: Bytes!
+
+  " The source of revenue (in alphabetical order) "
+  sources: [RevenueSource!]!
+
+  " The amount of revenue in USD (same order as sources) "
+  amountsUSD: [BigDecimal!]!
+}
+
+enum OracleSource {
+  UNISWAP
+  BALANCER
+  CHAINLINK
+  YEARN
+  SUSHISWAP
+  CURVE
+  ## Can add more
+}
+
+# Most lending protocols get token prices through onchain oracles.
+# As oracles are a source of truth it is important to understand the details.
+# This entity will help track oracle-related data in a market
+type Oracle @entity {
+  " { Market Address }-{ Token Address } "
+  id: Bytes!
```

Comment:
same as above, had no problems in comp v3 as well

Reply:
Since we are contacting two bytes, I blieve we are not inserting a "-" in between, right?

Reply:
The way I’ve been doing it for bridges was plain concatenation, without the `-`. When it renders as hex, the `-` will not show as `-`, so can be omitted IMO.

Only thing is we need to be careful with certain IDs. If it is concatenation of addresses, it’s fine. But if we concatenate say 2 arbitrary numbers we might end up with duplicate IDs:

For `{some num}{some other num}`:
- `11.concat(1)`
- `1.concat(11)`

It might be safer to have the `-` actually 🤔 
But then I think that `Bytes.fromUTF8(address-address) != Bytes.empty().concat(address).concat(“-“).concat(address)`. By not having the UTF8 char, it’s easier for people not to accidentaly generate IDs in different ways.

I think I’d leave the `-` out.

Reply:
fixed schema comments